docs/best-practices/java-coding-guidelines/readme.md

```diff
@@ -595,6 +594,20 @@ ImmutableList.Builder<String> l = new ImmutableList.Builder<>();
 ImmutableList.Builder<String> l = ImmutableList.builder();
 ```
 
+Guava collection factory methods such as `Maps.newHashMap`,
+`Lists.newArrayList` et al, should however be considered deprecated
+and should not be used.
+
+``` java
+// BAD. Don't do this.
+List<String> list = Lists.newArrayList();
+Map<String, String> map = Maps.newHashMap();
+
+// Good.
+List<String> list = new ArrayList<>();
+Map<String, String> map = new HashMap<>();
```

Comment:
I don't have super strong opinions here, at least in the basic zero-arg constructor case of `ArrayList`, `HashMap`, `LinkedHashMap`, `TreeMap`, `HashSet`, `LinkedHashSet`, and `TreeSet`. We could suggest replacing `Lists.newArrayList(E...)` with `Arrays.asList(E...)` as well.

The factory method `Lists.newArrayListWithCapacity(42)` is more descriptive than `new ArrayList<>(42)` constructor, but latter is more concise and has been around for 20+ years (excluding `<>` from Java 7 type inference).

I've definitely seen misuse `Lists.newArrayListWithExpectedSize(42)` where folks don't actually need/use the extra ~10% buffer allocated with that call, though there are occasional circumstances where one does actually want the buffer to avoid a full expansion & array copy.

If we do want to make these kind of recommendations, we should probably include error-prone checks to identify and generate suggested fixes, something like the below for `ArrayList`. The other `*List`, `*Map`, and `*Set` are left as an exercise for the reader 😄 

There's probably a separate discussion about baseline bumping [`JdkObsolete`](https://errorprone.info/bugpattern/JdkObsolete) to `ERROR` and killing off use of `LinkedList` in favor of `ArrayList` or `ArrayDeque` (if you need `Queue`/`Deque` access patterns) https://github.com/google/error-prone/issues/257 was a proposed error-prone check (though we could write our own here). See https://stuartmarks.wordpress.com/2015/12/18/some-java-list-benchmarks/ for some benchmarks (worth re-running on current CPU & memory systems).


```
package com.palantir.baseline.errorprone;

import com.google.auto.service.AutoService;
import com.google.errorprone.BugPattern;
import com.google.errorprone.BugPattern.SeverityLevel;
import com.google.errorprone.VisitorState;
import com.google.errorprone.bugpatterns.BugChecker;
import com.google.errorprone.fixes.SuggestedFix;
import com.google.errorprone.fixes.SuggestedFixes;
import com.google.errorprone.matchers.Description;
import com.google.errorprone.matchers.Matcher;
import com.google.errorprone.matchers.Matchers;
import com.google.errorprone.matchers.method.MethodMatchers;
import com.sun.source.tree.ExpressionTree;
import com.sun.source.tree.MethodInvocationTree;

@AutoService(BugChecker.class)
@BugPattern(
        name = "PreferCollectionConstructor",
        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
        linkType = BugPattern.LinkType.CUSTOM,
        severity = SeverityLevel.WARNING,
        summary = "Prefer ArrayList constructor instead of Guava's Lists.newArrayList* factory methods."
)
public final class PreferCollectionConstructor extends BugChecker
        implements BugChecker.MethodInvocationTreeMatcher {

    private static final long serialVersionUID = 1L;

    private static final Matcher<ExpressionTree> LISTS_NEW_ARRAY_LIST =
            MethodMatchers.staticMethod()
                    .onClass("com.google.common.collect.Lists")
                    .named("newArrayList")
                    .withParameters();

    private static final Matcher<ExpressionTree> LISTS_NEW_ARRAY_LIST_WITH_CAPACITY =
            MethodMatchers.staticMethod()
                    .onClass("com.google.common.collect.Lists")
                    .named("newArrayListWithCapacity")
                    .withParameters("int");

    private static final Matcher<ExpressionTree> LISTS_NEW_ARRAY_LIST_WITH_EXPECTED_SIZE =
            MethodMatchers.staticMethod()
                    .onClass("com.google.common.collect.Lists")
                    .named("newArrayListWithExpectedSize")
                    .withParameters("int");

    private static final Matcher<ExpressionTree> LISTS_FACTORY =
            Matchers.anyOf(LISTS_NEW_ARRAY_LIST, LISTS_NEW_ARRAY_LIST_WITH_CAPACITY);

    @Override
    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
        if (LISTS_FACTORY.matches(tree, state)) {
            SuggestedFix.Builder fix = SuggestedFix.builder();
            String qualifiedType = SuggestedFixes.qualifyType(state, fix, "java.util.ArrayList");
            String method = "new " + qualifiedType + "<>";
            return buildDescription(tree)
                    .setMessage("Prefer new ArrayList<> constructor")
                    .addFix(fix.replace(tree.getMethodSelect(), method).build())
                    .build();
        } else if (LISTS_NEW_ARRAY_LIST_WITH_EXPECTED_SIZE.matches(tree, state)) {
            // todo should we inline estimated sizing?
        }

        return Description.NO_MATCH;
    }

}
```

and tests:
```
/*
 * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.palantir.baseline.errorprone;

import com.google.errorprone.BugCheckerRefactoringTestHelper;
import com.google.errorprone.CompilationTestHelper;
import org.junit.Ignore;
import org.junit.Test;

public final class PreferCollectionConstructorTests {

    @Test
    public void should_not_use_Lists_newArrayList() {
        CompilationTestHelper.newInstance(PreferCollectionConstructor.class, getClass())
                .addSourceLines(
                        "Test.java",
                        "import com.google.common.collect.Lists;",
                        "import java.util.List;",
                        "class Test {",
                        "  List<?> f() {",
                        "    // BUG: Diagnostic contains: Prefer new ArrayList<>",
                        "    return Lists.newArrayList();",
                        "  }",
                        "}")
                .doTest();
    }

    @Test
    public void should_not_use_Lists_newArrayListWithCapacity() {
        CompilationTestHelper.newInstance(PreferCollectionConstructor.class, getClass())
                .addSourceLines(
                        "Test.java",
                        "import com.google.common.collect.Lists;",
                        "import java.util.List;",
                        "class Test {",
                        "  List<?> f() {",
                        "    // BUG: Diagnostic contains: Prefer new ArrayList<>",
                        "    return Lists.newArrayListWithCapacity(42);",
                        "  }",
                        "}")
                .doTest();
    }

    @Ignore
    @Test
    public void should_not_use_Lists_newArrayListWithExpectedSize() {
        CompilationTestHelper.newInstance(PreferCollectionConstructor.class, getClass())
                .addSourceLines(
                        "Test.java",
                        "import com.google.common.collect.Lists;",
                        "import java.util.List;",
                        "class Test {",
                        "  List<?> f() {",
                        "    // BUG: Diagnostic contains: Prefer new ArrayList<>",
                        "    return Lists.newArrayListWithExpectedSize(42);",
                        "  }",
                        "}")
                .doTest();
    }

    @Test
    public void should_use_new_ArrayList() {
        CompilationTestHelper.newInstance(PreferCollectionConstructor.class, getClass())
                .addSourceLines(
                        "Test.java",
                        "import java.util.ArrayList;",
                        "import java.util.List;",
                        "class Test {",
                        "  List<?> f() {",
                        "    return new ArrayList<>();",
                        "  }",
                        "}")
                .doTest();
    }

    @Test
    public void should_use_new_ArrayList_with_capacity() {
        CompilationTestHelper.newInstance(PreferCollectionConstructor.class, getClass())
                .addSourceLines(
                        "Test.java",
                        "import java.util.ArrayList;",
                        "import java.util.List;",
                        "class Test {",
                        "  List<?> f() {",
                        "    return new ArrayList<>(1);",
                        "  }",
                        "}")
                .doTest();
    }

    @Test
    public void auto_fix_Lists_newArrayList() {
        BugCheckerRefactoringTestHelper.newInstance(new PreferCollectionConstructor(), getClass())
                .addInputLines(
                        "Test.java",
                        "import com.google.common.collect.Lists;",
                        "import java.util.ArrayList;",
                        "import java.util.List;",
                        "class Test {",
                        "  List<?> f() {",
                        "    return Lists.newArrayList();",
                        "  }",
                        "}")
                .addOutputLines(
                        "Test.java",
                        "import com.google.common.collect.Lists;",
                        "import java.util.ArrayList;",
                        "import java.util.List;",
                        "class Test {",
                        "  List<?> f() {",
                        "    return new ArrayList<>();",
                        "  }",
                        "}")
                .doTest(BugCheckerRefactoringTestHelper.TestMode.TEXT_MATCH);
    }

    @Test
    public void auto_fix_Lists_newArrayListWithCapacity() {
        BugCheckerRefactoringTestHelper.newInstance(new PreferCollectionConstructor(), getClass())
                .addInputLines(
                        "Test.java",
                        "import com.google.common.collect.Lists;",
                        "import java.util.ArrayList;",
                        "import java.util.List;",
                        "class Test {",
                        "  List<?> f() {",
                        "    return Lists.newArrayListWithCapacity(4);",
                        "  }",
                        "}")
                .addOutputLines(
                        "Test.java",
                        "import com.google.common.collect.Lists;",
                        "import java.util.ArrayList;",
                        "import java.util.List;",
                        "class Test {",
                        "  List<?> f() {",
                        "    return new ArrayList<>(4);",
                        "  }",
                        "}")
                .doTest(BugCheckerRefactoringTestHelper.TestMode.TEXT_MATCH);
    }
}
```

Reply:
The javadoc for these methods state that they should be considered deprecated for code written in Java 7+. According to a discussion on the Guava lists they are not @deprecated for Java 6 compatibility. As of Guava 20.0 however, Java 6 is no longer supported. I expect these methods to be @deprecated at some point.

There's a longer internal Slack discussion about this. (Don't know if this repo is public so I'm not including a link to the discussion here though.)